### ISSUE TITLE
txt2card enhancement request

### ISSUE THREAD

**Issue by wpappas** [SDL-Hercules-390/hyperion#685]
_**LOW PRIORITY**_

It would be nice if `txt2card` was modified to allow its output to be piped into another process.

_**Rationale:**_

I am using the `sockdev` option for a 3505 reader on Hercules, and I sometimes want to send ASCII decks to it, but other times want to use it with EBCDIC.

If I sent up the reader for EBCDIC, I can use `txt2card` to convert an ASCII deck to EBCDIC, and then send that, but it would be nice if I could pipe the output of `txt2card` directly into `netcat`, thus avoiding creating a temporary EBCDIC version of it.

I've created a bit of a hack workaround using:
```
 txt2card default $1 /dev/stdout | tail -n +6 | nc -w1 localhost 3505
```
But it is a hack, as it relies on the status output of `txt2card` being exactly 6 lines.

_**LOW PRIORITY**_

**Comment by Fish-Git** [comment]
Bill,

I haven't personally tried it _(I myself use HercRdr on Windows)_, but wouldn't:
```
txt2card default $1 $1.temp && nc -w1 localhost 3505 < $1.temp && rm $1.temp
```
work just as well?

**Comment by Fish-Git** [comment]
Or is your primary concern (wish) to simply AVOID creating a temporary file altogether?

**Comment by Fish-Git** [comment]
I suppose I could modify the `outfile` 3rd argument to accept a port number instead of a filename if it begins with, say, a colon. The results (status output) would still be written to stdout/stderr as usual, but the translated file `outfile` , instead of being written as a regular file (as it does today), would instead be written to the specified port. That would allow you to do everything with one simple command:

       txt2card default infile :3505

Would that work?

**Comment by wpappas** [comment]
> I suppose I could modify the `outfile` 3rd argument to accept a port number instead of a filename if it begins with, say, a colon. The results (status output) would still be written to stdout/stderr as usual, but the translated file `outfile` , instead of being written as a regular file (as it does today), would instead be written to the specified port. That would allow you to do everything with one simple command:
> 
> ```
>    txt2card default infile :3505
> ```
> 
> Would that work?

I would like to avoid the creation of the temporary file.  As I thought more about it, a real easy answer would be to either:
A) Add a quiet option that inhibits any of the status output
B) Write all the status output to stderr

Your answer of adding a port would also work, but that may be more effort than this deserves.
Make sense?

--Bill

**Comment by Fish-Git** [comment]
> As I thought more about it, a real easy answer would be to either:
> 
> 1. Add a quiet option that inhibits any of the status output

I'll see what I can do, but it might not be possible. All Hercules utilities always display a standard startup banner:
```
HHC02499I Hercules utility txt2card - Translate text file to EBCDIC utility - version 4.8.0.11309-SDL-DEV-gd7740892
HHC01414I (C) Copyright 1999-2024 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SDL 4.x Hyperion version of Hercules **
HHC01415I Build date: Sep 12 2024 at 00:36:31
```
as soon as they start, _before_ parsing any command-line arguments.

> 2\. Write all the status output to stderr

That would break standard command-line tool/utility behavior. And by that, I mean ALL command-line tool/utility behavior, not just Hercules. Standard behavior for any/all command-line tools/utilities on virtuall _all_ desktop operating systems (Windows and Linux alike) is to issue normal status/progress (informational) messages to stdout, and error messages to stderr.  That way users can log each separately or just one or the other to a file, etc: `command > stdout.log 2>stderr.log`, or `command 2>errors-only.log`, etc.

> Your answer of adding a port would also work, but that may be more effort than this deserves.

Not at all! It should be relatively simple and straghtfoward IMO. I _did_ write the Windows HercRdr tool after all, which already does the same thing. So all the code I need is already written. The only thing my Windows HercRdr tool currently _doesn't_ do (that `txt2card` does) is the ASCII-->EBCDIC translation side of it _(which, now that I think abot it, would be a handy optional feature to add to it!)_.

Your original request/suggestion is a good one, Bill, so I don't mind at all specnding a _little_ bit of time and effort to make a silly/handy tool/utility like `txt2card` even better.

It'll be fun to see how quickly I can get it done. &nbsp;`:)`

**Comment by peter-sylvester** [comment]
iconv may be an alternative.

**Comment by wpappas** [comment]
> iconv may be an alternative.

Hi Peter,
I have messed with iconv in the past and couldn't get it to work and then stumbled across txt2card (which does do the proper conversion).  I just tried using iconv again using:
iconv -f ASCII -t EBCDIC infile.jcl > outfile.card
and it errors out with "iconv: failed to start conversion processing".

Not sure what's going on with that, but it is unrelated to Hercules so not sure it makes sense to discuss it more here.

--Bill

**Comment by Fish-Git** [comment]
> Looks like it might be a padding issue.

To prove it, try ensuring that either: a) each line of your ASCII file is exactly 80 characters long (i.e. blank pad each line to 80 characters), and/or b) that NO newlines are used AT ALL.

`txt2card` begins a new 80-byte card whenever it reads a newline, whereas I'm guessing `iconv` simply translates each character of the file as-is (which would indeed confuse Hercules if the size of the file wasn't an exact multiple of 80 bytes).

**Comment by peter-sylvester** [comment]
iconv has no idea about the length of a "card". Adding the following script to the pipe before iconv may help.  

#!/usr/bin/python
import sys
for line in sys.stdin:
    print( '{: <80}'.format(line)

**Comment by Fish-Git** [comment]
> `iconv` has no idea about the length of a "card".

Duh! No shit. That's precisely the point I was trying to make.

> Adding the following script to the pipe before iconv may help:

Which is essentially the automated way of doing what I suggested trying.

**Comment by wpappas** [comment]
> > `iconv` has no idea about the length of a "card".
> 
> That's precisely the point I was trying to make.
> 
> > Adding the following script to the pipe before iconv may help:
> 
> Which is essentially the automated way of doing what I suggested trying.

Yes -- padding confirmed.  I haven't finished testing it, but the following script basically mimics txt2card using iconv:

#!/bin/bash
ENCODING=${2:-ASCII}

function padfile()
{
    IFS_SAVE="${IFS}"
    IFS=''
    while read -r line; do
        printf %-80s ${line}
    done < ${1}
    IFS="${IFS_SAVE}"
}
padfile ${1} | iconv -f ${ENCODING} -t EBCDIC-CP-US -

**Comment by peter-sylvester** [comment]
hercifc writes the banner to stdout, not to stderr.
See the end of the hscutl routine initialize_utility.
Seems not to be rocket science to have parameter or for ALL utilities that
either set a quiet mode or stderr. I think all utilities would benefit from that, e.g. dasdls  

If the script about is considered a correct implementation of whatever specification, then this does not seem to be the case for txt2card.  Try to use "EXTERNALGUI" as output filename for example.  

best

**Comment by Fish-Git** [comment]
> `hercifc` writes the banner to stderr, not to stdout.

`hercifc` is a special case. It is not meant to be run interactivly. It's designed to be run interally, with stdout piped back to the Hercules that started it (so it can know whether the commands it sends it were successful or not).

All other utilities (like `txt2card`) are interactive. They are started manually by the user from the command-line.

> Seems not to be rocket science to have parameter or for ALL utilities that either set a quiet mode or stderr. I think all utilities would benefit from that, e.g. `dasdls`.

No, you're right. It's certainly not rocket science. Such could easily(?) be done with a little bit of effort. But the question is why? Why change each utility's existing behavior just because someone reported an issue with a single specific utility?

> If the script about is considered a correct implementation of whatever specification, then this does not seem to be the case for `txt2card`.

It's not a specification or a standard. It's simply a widespread common convention.

> Try to use `EXTERNALGUI` as output filename for example.

Why tells me, by virtue of your having mentioning it, that you are already aware of how that undocumented special argument is handled by Hercules (as am I, since I'm the one that wrote it!), so why even bring it up?

Unless you're implying that it should be documented somewhere as being an internal Hercules "reserved" argument that when specified as the last argument triggers special external GUI behavior?

Which I would respond with the fact that no one that I know of has ever complained or tripped over it in all the years since Hercules has supported external GUIs (which has been virtually forever, i.e. over 20+ years I think).

**Comment by Fish-Git** [comment]
> I suppose I could modify the `outfile` 3rd argument to accept a port number instead of a filename if it begins with, say, a colon. The results (status output) would still be written to stdout/stderr as usual, but the translated file `outfile` , instead of being written as a regular file (as it does today), would instead be written to the specified port. That would allow you to do everything with one simple command:
> 
> ```
>    txt2card default infile :3505
> ```

### DONE!

Fixed by commit baac8e939297f38d15fee55ae5096a850edffb07.

Sorry it took me so long. &nbsp;See <http://www.softdevlabs.com/news.html> for the reason why.

Please also note that I did make one small change to the syntax:

```
     txt2card  codepage  ifile  {ofile | host:port}
```

That is to say, the 3rd option must either be a filename or a `hostname:port`, not just a port number. (i.e. the hostname (or IP number) is required, and must be followed by a colon `:` and port number.)

I would also like to remind others who may wish to utilize this new ability of `txt2card` that the device statement for your Hercules "Hot Reader" device _must_ specify the `ebcdic` option too, not the `ascii` option. E.g.:

<pre>
     000C    3505    192.168.0.107:3505  sockdev  <b>ebcdic</b>  autopad  eof
</pre>
